ffi/derive/src/convert.rs

```diff
@@ -32,105 +39,66 @@ pub fn derive_ffi_type(input: DeriveInput) -> TokenStream2 {
-        syn::Data::Union(item) => abort!(item.union_token, "Unions are not supported"),
-    };
-
-    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
-
-    quote! {
-        unsafe impl #impl_generics iroha_ffi::ir::Transmute for #name #ty_generics #where_clause {
-            type Target = #inner;
-
-            // FIXME: We should force the transparent type to have `from_inner` and derive implementation
-            // only if requested specifically by the user (via macro attribute). The reason for this is
-            // that even though two types are equal on the byte level, they may have different semantics
-            // that cannot be proven when deriving `from_inner` (e.g. `NonZeroU8` is a wrapper for `u8`)
-            // Deriving implementation of this method cannot be guaranteed to always return true
-            unsafe fn is_valid(inner: &#inner) -> bool {
-                true
-            }
+            derive_ffi_type_for_struct(name, input.generics, &item)
         }
+        Data::Union(item) => {
+            if !is_non_owning_item(
+                &input.attrs,
+                item.fields.named.iter().map(|field| &field.ty),
+            ) {
+                abort!(name, "if applicable, attach `#[ffi_type(non_owning)`")
+            }
 
-        impl #impl_generics iroha_ffi::ir::Ir for #name #ty_generics #where_clause {
-            type Type = iroha_ffi::ir::Transparent<Self>;
-        }
-        impl #impl_generics iroha_ffi::ir::Ir for &#name #ty_generics #where_clause {
-            type Type = iroha_ffi::ir::Transparent<Self>;
+            derive_ffi_type_for_union(name, input.generics, &item)
         }
     }
 }
 
-fn derive_ffi_type_for_opaque_struct(name: &Ident, mut generics: Generics) -> TokenStream2 {
-    let lifetime = quote!('__iroha_ffi_itm);
+// TODO: The type is also non-owning if it implements `Copy`
+fn is_non_owning_item<'itm>(
+    attrs: &[syn::Attribute],
+    fields: impl IntoIterator<Item = &'itm syn::Type>,
+) -> bool {
+    let non_owning: syn::Attribute = parse_quote! {#[ffi_type(non_owning)]};
 
-    let (impl_generics, ty_generics, where_clause) =
-        split_for_impl_with_type_params(&mut generics, &[]);
+    if attrs.iter().any(|attr| *attr == non_owning) {
+        return true;
+    }
+    if fields.into_iter().all(|field_ty| !is_owning_type(field_ty)) {
+        return true;
+    }
 
-    quote! {
-        unsafe impl<#lifetime, #impl_generics> iroha_ffi::ir::Transmute for &#lifetime #name #ty_generics #where_clause {
-            type Target = *const #name #ty_generics;
+    false
+}
 
-            unsafe fn is_valid(source: &Self::Target) -> bool {
-                source.as_ref().is_some()
-            }
-        }
-        unsafe impl<#lifetime, #impl_generics> iroha_ffi::ir::Transmute for &#lifetime mut #name #ty_generics #where_clause {
-            type Target = *mut #name #ty_generics;
+fn is_owning_type(type_: &syn::Type) -> bool {
```

Comment:
By `owning` we mean do not reference local context, right?

Reply:
whoops, this might have been a leftover

Reply:
I think it's more clear now